stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,855 @@
+   *   not less than t0.
+   * @param rel_tol_f Relative tolerance for forward problem passed to CVODES
+   * @param abs_tol_f Absolute tolerance for forward problem passed to CVODES
+   * @param rel_tol_b Relative tolerance for backward problem passed to CVODES
+   * @param abs_tol_b Absolute tolerance for backward problem passed to CVODES
+   * @param rel_tol_q Relative tolerance for quadrature problem passed to CVODES
+   * @param abs_tol_q Absolute tolerance for quadrature problem passed to CVODES
+   * @param max_num_steps Upper limit on the number of integration steps to
+   *   take between each output (error if exceeded)
+   * @param num_checkpoints Number of integrator steps after which a checkpoint
+   * is stored for the backward pass
+   * @param interpolation_polynomial type of polynomial used for interpolation
+   * @param solver_f solver used for forward pass
+   * @param solver_b solver used for backward pass
+   *   take between each output (error if exceeded)
+   * @param[in, out] msgs the print stream for warning messages
+   * @param args Extra arguments passed unmodified through to ODE right hand
+   * side function
+   * @return Solution to ODE at times \p ts
+   * @return a vector of states, each state being a vector of the
+   * same size as the state variable, corresponding to a time in ts.
+   */
+  template <require_eigen_col_vector_t<T_y0>* = nullptr>
+  cvodes_integrator_adjoint_vari(
+      const char* function_name, const F& f, const T_y0& y0, const T_t0& t0,
+      const std::vector<T_ts>& ts, double rel_tol_f, Eigen::VectorXd abs_tol_f,
+      double rel_tol_b, Eigen::VectorXd abs_tol_b, double rel_tol_q,
+      double abs_tol_q, long int max_num_steps, long int num_checkpoints,
+      int interpolation_polynomial, int solver_f, int solver_b,
+      std::ostream* msgs, const T_Args&... args)
+      : function_name_(function_name),
+        vari(NOT_A_NUMBER),
+        N_(y0.size()),
+        returned_(false),
+        memory(NULL),
+        rel_tol_f_(rel_tol_f),
+        abs_tol_f_(abs_tol_f),
+        rel_tol_b_(rel_tol_b),
+        abs_tol_b_(abs_tol_b),
+        rel_tol_q_(rel_tol_q),
+        abs_tol_q_(abs_tol_q),
+        max_num_steps_(max_num_steps),
+        num_checkpoints_(num_checkpoints),
+        interpolation_polynomial_(interpolation_polynomial),
+        solver_f_(solver_f),
+        solver_b_(solver_b),
+        lmm_f_(solver_f_ % 2 ? CV_ADAMS : CV_BDF),
+        lmm_b_(solver_b_ % 2 ? CV_ADAMS : CV_BDF),
+        msgs_(msgs),
+
+        t0_vars_(count_vars(t0)),
+        ts_vars_(count_vars(ts)),
+        y0_vars_(count_vars(y0)),
+        args_vars_(count_vars(args...)),
+        t0_varis_(
+            ChainableStack::instance_->memalloc_.alloc_array<vari*>(t0_vars_)),
+        ts_varis_(
+            ChainableStack::instance_->memalloc_.alloc_array<vari*>(ts_vars_)),
+        y0_varis_(
+            ChainableStack::instance_->memalloc_.alloc_array<vari*>(y0_vars_)),
```

Comment:
I'm not sure I understand what's happening here, can all of these be deeply nested containers of containers like `std::vector<std::vector<...>>`? There should be docs specifying what the actual acceptable types are for these parameters

Reply:
Sure. will add the doc.

t0: var or double
ts: var or double (resulting in a std:vector of that)
y0: Eigen vector of double or var

Reply:
Regarding:

> Why does cvodes_integrator_adjoint_memory exist?
> I don't like the private friend class pattern

Not sure. Likely just the first way I got things working and things can be cleaner now with make_chainable_ptr and whatnot.

Reply:
Okay cool cool I think I can zip all this up then I just didn't know of it was there for reuse or something

Reply:
We definetly do not need two classes, I think. However, what would be nice to retain is a callback class for CVODES which is templated for the ODE RHS function. The reason is that we can then provide specializations of that class in order to use symbolic math which would need to be pre-generated as C++ extra code. This can significantly speed up large ODE systems. I don't intend to make this a requirement here, but if the design allows for that then it would be great.